### PR TITLE
Specialise overloaded functions to Graph Int

### PR DESCRIPTION
**Pull request by nobrakal**
Here the code for #75 

I can add some benchs if you want :)

### CODE REVIEW

**Comment by snowleopard** [comment]
Thanks! Yes, it would be nice to see some benchmarks. I'm not sure if it takes a lot of time for you -- ideally we should automate this somehow, creating a regression suite to see performance effects of all PRs. Do you think you could implement something like this?

File: src/Algebra/Graph.hs [hunk 1]
```diff
@@ -395,6 +395,8 @@ foldg e v o c = go
 isSubgraphOf :: Ord a => Graph a -> Graph a -> Bool
 isSubgraphOf x y = overlay x y == y
 
+{-# SPECIALISE isSubgraphOf :: Graph Int -> Graph Int -> Bool #-}
```
Comment (snowleopard) [inline src/Algebra/Graph.hs]: Can you skip the empty line above? It's not immediately obvious to which function the specialisations apply.

Comment (nobrakal) [reply src/Algebra/Graph.hs]: Of course, I thought that rendered not so well, but you are right

Comment (nobrakal) [reply src/Algebra/Graph.hs]: In fact, because it looks like a type definition, maybe adding it just below the actual type definition is a good idea

Comment (snowleopard) [reply src/Algebra/Graph.hs]: Another option is to group all these pragmas together, say at the bottom of the file with a comment explaining what it achieves.

**Comment by nobrakal** [comment]
Mmhh that can certainly be done. I did not think about it. A bash-script with ~ 10 lines can do the trick

**Comment by snowleopard** [comment]
> A bash-script with ~ 10 lines can do the trick

Yep, and we should call this script from our CI scripts.

File: src/Algebra/Graph.hs (hunk 1 above)
Comment (nobrakal) [reply src/Algebra/Graph.hs]: Yes, but one can forget to check this. I like that when I am reading the code and I have everything under my eyes

Comment (snowleopard) [reply src/Algebra/Graph.hs]: OK, let's place it next to function definitions. Perhaps, below the comment but on top of function type signature? Unless this messes up Haddock generation.

Comment (nobrakal) [reply src/Algebra/Graph.hs]: Ok I will check this

**Comment by nobrakal** [comment]
Hop, a script is living here: https://github.com/nobrakal/benchAlgaPr

It will bench a directory containing a different version of alga again `alga/master/HEAD`. Fore example here the content of `res.md` on this PR, using
```Bash
./compare.sh ../alga res.md hasVertex removeVertex hasEdge removeEdge
```


Using [("Mesh",3),("Clique",3)] as graphs

## hasVertex

Descritpion: Test if the given vertex is in the graph

### Mesh
<TABLE>
   <TR>
      <TH>
      </TH>
      <TH CLASS = "thinright">
         1
      </TH>
      <TH CLASS = "thinright">
         10
      </TH>
      <TH>
         100
      </TH>
   </TR>
   <TR>
      <TH>
         Alga
      </TH>
      <TD CLASS = "thinright">
         24.75 ns
      </TD>
      <TD CLASS = "thinright">
         81.59 ns
      </TD>
      <TD>
         81.29 ns
      </TD>
   </TR>
   <TR>
      <TH>
         AlgaOld
      </TH>
      <TD CLASS = "thinright">
         31.46 ns
      </TD>
      <TD CLASS = "thinright">
         114.3 ns
      </TD>
      <TD>
         112.3 ns
      </TD>
   </TR>
</TABLE>

### Clique
<TABLE>
   <TR>
      <TH>
      </TH>
      <TH CLASS = "thinright">
         1
      </TH>
      <TH CLASS = "thinright">
         10
      </TH>
      <TH>
         100
      </TH>
   </TR>
   <TR>
      <TH>
         Alga
      </TH>
      <TD CLASS = "thinright">
         25.27 ns
      </TD>
      <TD CLASS = "thinright">
         396.2 ns
      </TD>
      <TD>
         37.96 μs
      </TD>
   </TR>
   <TR>
      <TH>
         AlgaOld
      </TH>
      <TD CLASS = "thinright">
         31.52 ns
      </TD>
      <TD CLASS = "thinright">
         590.3 ns
      </TD>
      <TD>
         55.66 μs
      </TD>
   </TR>
</TABLE>


SUMMARY:

 * Alga was the fastest 16 times


ABSTRACT:
(Based on an average of the ratio between largest benchmarks)

 * Alga was 1.36 times faster than AlgaOld

## hasEdge

Descritpion: Test if the given edge is in the graph (with arguments both in the graph and not in the graph (where applicable))

### Mesh
<TABLE>
   <TR>
      <TH>
      </TH>
      <TH CLASS = "thinright">
         1
      </TH>
      <TH CLASS = "thinright">
         10
      </TH>
      <TH>
         100
      </TH>
   </TR>
   <TR>
      <TH>
         Alga
      </TH>
      <TD CLASS = "thinright">
         343.6 ns
      </TD>
      <TD CLASS = "thinright">
         1.332 μs
      </TD>
      <TD>
         1.339 μs
      </TD>
   </TR>
   <TR>
      <TH>
         AlgaOld
      </TH>
      <TD CLASS = "thinright">
         552.4 ns
      </TD>
      <TD CLASS = "thinright">
         2.292 μs
      </TD>
      <TD>
         2.280 μs
      </TD>
   </TR>
</TABLE>

### Clique
<TABLE>
   <TR>
      <TH>
      </TH>
      <TH CLASS = "thinright">
         1
      </TH>
      <TH CLASS = "thinright">
         10
      </TH>
      <TH>
         100
      </TH>
   </TR>
   <TR>
      <TH>
         Alga
      </TH>
      <TD CLASS = "thinright">
         345.2 ns
      </TD>
      <TD CLASS = "thinright">
         6.372 μs
      </TD>
      <TD>
         573.9 μs
      </TD>
   </TR>
   <TR>
      <TH>
         AlgaOld
      </TH>
      <TD CLASS = "thinright">
         543.3 ns
      </TD>
      <TD CLASS = "thinright">
         7.943 μs
      </TD>
      <TD>
         637.9 μs
      </TD>
   </TR>
</TABLE>


SUMMARY:

 * Alga was the fastest 18 times


ABSTRACT:
(Based on an average of the ratio between largest benchmarks)

 * Alga was 1.42 times faster than AlgaOld

## removeVertex

Descritpion: Remove a vertex of the graph

### Mesh
<TABLE>
   <TR>
      <TH>
      </TH>
      <TH CLASS = "thinright">
         1
      </TH>
      <TH CLASS = "thinright">
         10
      </TH>
      <TH>
         100
      </TH>
   </TR>
   <TR>
      <TH>
         Alga
      </TH>
      <TD CLASS = "thinright">
         36.52 ns
      </TD>
      <TD CLASS = "thinright">
         339.7 ns
      </TD>
      <TD>
         325.5 ns
      </TD>
   </TR>
   <TR>
      <TH>
         AlgaOld
      </TH>
      <TD CLASS = "thinright">
         35.05 ns
      </TD>
      <TD CLASS = "thinright">
         327.6 ns
      </TD>
      <TD>
         337.4 ns
      </TD>
   </TR>
</TABLE>

### Clique
<TABLE>
   <TR>
      <TH>
      </TH>
      <TH CLASS = "thinright">
         1
      </TH>
      <TH CLASS = "thinright">
         10
      </TH>
      <TH>
         100
      </TH>
   </TR>
   <TR>
      <TH>
         Alga
      </TH>
      <TD CLASS = "thinright">
         36.51 ns
      </TD>
      <TD CLASS = "thinright">
         3.067 μs
      </TD>
      <TD>
         440.7 μs
      </TD>
   </TR>
   <TR>
      <TH>
         AlgaOld
      </TH>
      <TD CLASS = "thinright">
         35.28 ns
      </TD>
      <TD CLASS = "thinright">
         3.007 μs
      </TD>
      <TD>
         422.8 μs
      </TD>
   </TR>
</TABLE>


SUMMARY:

There was 6 ex-aequo


ABSTRACT:
(Based on an average of the ratio between largest benchmarks)

 * AlgaOld was 1.00 times faster than Alga

## removeEdge

Descritpion: Remove an edge of the graph

### Mesh
<TABLE>
   <TR>
      <TH>
      </TH>
      <TH CLASS = "thinright">
         10
      </TH>
      <TH>
         100
      </TH>
   </TR>
   <TR>
      <TH>
         Alga
      </TH>
      <TD CLASS = "thinright">
         2.001 μs
      </TD>
      <TD>
         2.008 μs
      </TD>
   </TR>
   <TR>
      <TH>
         AlgaOld
      </TH>
      <TD CLASS = "thinright">
         2.204 μs
      </TD>
      <TD>
         2.211 μs
      </TD>
   </TR>
</TABLE>

### Clique
<TABLE>
   <TR>
      <TH>
      </TH>
      <TH CLASS = "thinright">
         10
      </TH>
      <TH>
         100
      </TH>
   </TR>
   <TR>
      <TH>
         Alga
      </TH>
      <TD CLASS = "thinright">
         15.67 μs
      </TD>
      <TD>
         7.655 ms
      </TD>
   </TR>
   <TR>
      <TH>
         AlgaOld
      </TH>
      <TD CLASS = "thinright">
         17.04 μs
      </TD>
      <TD>
         7.907 ms
      </TD>
   </TR>
</TABLE>


SUMMARY:

 * Alga was the fastest 4 times
There was 8 ex-aequo


ABSTRACT:
(Based on an average of the ratio between largest benchmarks)

 * Alga was 1.07 times faster than AlgaOld


The specialization seems to be useful only on graph inspection...

**Comment by snowleopard** [comment]
@nobrakal Great! Could you send a separate PR to invoke this script from our CI scripts?

**Comment by nobrakal** [comment]
For CI I don't know if this is a good idea.

It's output is not very good to work with, and because I cannot say which benchmark is good to do, it will run the whole bench-suite, producing many unrelevant output.

**Comment by snowleopard** [comment]
> It's output is not very good to work with

Yes, right now it's difficult to view without rendering in a Markdown viewer, but still it's better than nothing. As a maintainer of Alga I will be able to review the performance effect of any PR, which is very important. Otherwise, it's really easy to accidentally introduce a performance regression.

Perhaps, in future we can add an option to render the output in a more terminal friendly way. It's probably not even too difficult, we just need to redirect output to a Markdown file and then run an existing text-only renderer for Markdown files (right in the CI script).

> I cannot say which benchmark is good to do, it will run the whole bench-suite

Yes, that's the only way. It's the same with the testsuite really -- we always run all tests even if we change a comment, because there is no easy way to do better.

**Comment by nobrakal** [comment]
> Perhaps, in future we can add an option to render the output in a more terminal friendly way. 

It can even be done directly in the bench suite, only require some tweaks

> As a maintainer of Alga I will be able to review the performance effect of any PR, which is very important. 

Ok, so I will send a PR. Since it is only for a quick review, do you only want to "Best,Abstract" part, and not the detail ? It will be simpler to check I think

**Comment by snowleopard** [comment]
@nobrakal Thanks! 

I've merged this PR.

**Pull request merged**